### PR TITLE
Adds Projection, Measurement Classes to PennyLane

### PR DESCRIPTION
**Pull request by Lucaman99**
**Context:**

Part of the effort to add conditional operations into PennyLane as outlined in the corresponding [ADR](https://github.com/PennyLaneAI/adrs/blob/master/documents/041-conditional_operations.rst).

**Description of the Change:**

Adds basic classes for projection operations, and measurements in the computational basis.
**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/operation.py [hunk 1]
```diff
@@ -964,6 +964,75 @@ def kraus_matrices(self):
...
+
+    * :attr:`~.Projection._projectors`
+
+    Args:
+        params (tuple[float, int, array, Variable]): operation parameters
+
+    Keyword Args:
+        wires (Sequence[int]): Subsystems the projection acts on. If not given, args[-1]
+            is interpreted as wires.
+        do_queue (bool): Indicates whether the operation should be
```
Comment (ixfoduap) [inline pennylane/operation.py:988]: Does `do_queue` also appear in other operations? Seems new to me, but I guess I missed those changes

File: pennylane/ops/qubit.py [hunk 2]
```diff
@@ -1740,6 +1740,39 @@ def diagonalizing_gates(self):
...
+    num_params = 0
+    num_wires = AnyWires
+    par_domain = None
+
+    @classmethod
+    def _projectors(cls, wires):
+        nr_proj = int(2 ** len(wires))
+        projectors = [np.zeros((nr_proj, nr_proj)) for i in range(nr_proj)]
+
+        for i in range(nr_proj):
```
Comment (ixfoduap) [inline pennylane/ops/qubit.py:1770]: So elegant! :star_struck:

File: tests/test_operation.py [hunk 3]
```diff
@@ -1471,3 +1471,36 @@ def test_cry(self):
             ]
         )
         assert np.allclose(derivative, expected_derivative)
+
+
+class TestProjector:
```
Comment (ixfoduap) [inline tests/test_operation.py:1476]: Great test! :star2:

File: tests/test_operation.py [hunk 4]
```diff
@@ -1471,3 +1471,36 @@ def test_cry(self):
...
+        expected = [
+            np.array([[0.15, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]),
+            np.array([[0, 0, 0, 0], [0, 0.15, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]),
+            np.array([[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0.15, 0], [0, 0, 0, 0]]),
+            np.array([[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0.15]])
+        ]
+        proj = DummyProjector(0.5, 0.3, wires=["a", 2])
+
+        res = [a == b for a, b in zip(expected, proj.projectors)]
+        assert np.all(res)
```
Comment (ixfoduap) [inline tests/test_operation.py:1506]: As per codecov warning, just needs to add a test that a warning is issued when the projectors are not implemented

**Review by ixfoduap — APPROVED** [review]
Look great! Just needs fixing of tests not passing

File: pennylane/operation.py (hunk 1 above)
Comment (Lucaman99) [reply pennylane/operation.py:988]: I believe it does, but I'll check again!

File: pennylane/operation.py [hunk 5]
```diff
@@ -964,6 +964,77 @@ def kraus_matrices(self):
         return self._kraus_matrices(*self.parameters)
 
 
+class Projection(Operation):
```
Comment (trbromley) [inline pennylane/operation.py:967]: Out of curiosity, what is the motivation to inherit from `Operation` rather than `Operator`? Do we plan on using some of the methods unique to `Operation` such as `generator` or `decomposition`? Do these make sense for a projection?

File: pennylane/operation.py [hunk 6]
```diff
@@ -964,6 +964,77 @@ def kraus_matrices(self):
...
+    """Base class for projection operations
```
Comment (trbromley) [inline pennylane/operation.py:968]:
```suggestion
    """Base class for projection operations.
```

File: pennylane/operation.py [hunk 7]
```diff
@@ -964,6 +964,77 @@ def kraus_matrices(self):
...
+
+    As with :class:`~.Operation`, the following class attributes must be
+    defined for all projections:
+
+    * :attr:`~.Operator.num_params`
```
Comment (trbromley) [inline pennylane/operation.py:973]: Another curiosity - do we expect `Projection` to have more than 0 params? I guess you could feed the diagonalizing unitary?

File: pennylane/operation.py [hunk 8]
```diff
@@ -964,6 +964,77 @@ def kraus_matrices(self):
...
+    To define a projection, the following attribute of :class:`~.Projection`
+    can be used to list the corresponding projectors.
+
+    * :attr:`~.Projection._projectors`
+
+    Args:
+        params (tuple[float, int, array, Variable]): operation parameters
+
+    Keyword Args:
+        wires (Sequence[int]): Subsystems the projection acts on. If not given, args[-1]
```
Comment (trbromley) [inline pennylane/operation.py:986]:
```suggestion
        wires (Iterable[Number, str], Number, str, Wires): Subsystems the projection acts on. If not given, args[-1]
```
This is to support the `Wires` class

File: pennylane/operation.py [hunk 9]
```diff
@@ -964,6 +964,77 @@ def kraus_matrices(self):
...
+    * :attr:`~.Operator.num_wires`
+    * :attr:`~.Operator.par_domain`
+
+    To define a projection, the following attribute of :class:`~.Projection`
+    can be used to list the corresponding projectors.
+
+    * :attr:`~.Projection._projectors`
+
+    Args:
+        params (tuple[float, int, array, Variable]): operation parameters
```
Comment (trbromley) [inline pennylane/operation.py:983]:
```suggestion
        params (tuple[float, int, array]): operation parameters
```
Maybe @josh146 has some thoughts, but since we're moving away from the old `Variable` class, this is probably no longer needed (also applies to other cases like in `Operation`, so we could leave this for a follow up)

File: pennylane/operation.py (hunk 1 above)
Comment (trbromley) [reply pennylane/operation.py:988]: @josh146 does `do_queue` still apply to the new QNode/tape core?

File: pennylane/operation.py [hunk 10]
```diff
@@ -964,6 +964,77 @@ def kraus_matrices(self):
...
+    Keyword Args:
+        wires (Sequence[int]): Subsystems the projection acts on. If not given, args[-1]
+            is interpreted as wires.
+        do_queue (bool): Indicates whether the operation should be
+            immediately pushed into a :class:`BaseQNode` circuit queue.
+            This flag is useful if there is some reason to run an Operation
+            outside of a BaseQNode context.
+    """
+
+    # pylint: disable=abstract-method
```
Comment (trbromley) [inline pennylane/operation.py:994]: Do we need this, since we don't have any abstract methods?
On the other hand _should_ we have abstract methods?
`Channel` does declare `_kraus_matrices` as an abstract method, but `Operator` doesn't have `_matrix` as an abstract method, so we're not consistent :thinking: 

Maybe it's fine here to keep as is (just remove the pylint ignore - unless I missed something!)

File: pennylane/operation.py [hunk 11]
```diff
@@ -964,6 +964,77 @@ def kraus_matrices(self):
...
+
+    @classmethod
+    def _projectors(cls, *params, wires):
+        r"""Projectors representing a projection operation, specified in the
+        computational basis.
+
+        This is a class method that should be defined for all
+        new projections. It returns the projectors representing
+        the projection in the computational basis.
```
Comment (trbromley) [inline pennylane/operation.py:1003]:
```suggestion
        new projections.
```
Thought this was a bit similar to above

File: pennylane/operation.py [hunk 12]
```diff
@@ -964,6 +964,77 @@ def kraus_matrices(self):
...
+
+        This private method allows projectors to be computed
+        directly without instantiating the projection first.
+
+        **Example**
+
+        >>> qml.Measure(wires=0)
```
Comment (trbromley) [inline pennylane/operation.py:1010]:
```suggestion
        >>> qml.Measure._projectors(wires=0)
```

File: pennylane/operation.py [hunk 13]
```diff
@@ -964,6 +964,77 @@ def kraus_matrices(self):
         return self._kraus_matrices(*self.parameters)
```
Comment (trbromley) [inline pennylane/operation.py:964]: Worth adding to the docstrings at the top of the file? These actually are rendered in the main docs:
https://pennylane.readthedocs.io/en/latest/code/qml_operation.html

Comment (trbromley) [reply pennylane/operation.py:964]: But not too high priority, if anything a sentence or two

File: pennylane/ops/qubit.py [hunk 14]
```diff
@@ -1882,6 +1882,39 @@ def diagonalizing_gates(self):
         return [QubitUnitary(self.eigendecomposition["eigvec"].conj().T, wires=list(self.wires))]
 
 
+# =============================================================================
```
Comment (trbromley) [inline pennylane/ops/qubit.py:1885]: Worth adding a new section below https://pennylane.readthedocs.io/en/latest/introduction/operations.html#qubit-observables for measurements?

E.g. "Qubit Measurements"

File: pennylane/ops/qubit.py [hunk 15]
```diff
@@ -1882,6 +1882,39 @@ def diagonalizing_gates(self):
...
+    A measurement in the computational basis.
+
+    **Details:**
+
+    * Number of wires: Any
+    * Number of parameters: 0
+    * Gradient recipe: None
+
+    Args:
+        wires (Sequence[int] or int): the wire(s) the operation acts on
```
Comment (trbromley) [inline pennylane/ops/qubit.py:1901]: Perhaps we could add an example usage below here? It'll help users understand that this is for measurements partway through the circuit

File: pennylane/ops/qubit.py (hunk 14 above)
Comment (trbromley) [inline pennylane/ops/qubit.py:1885]: Not a major thing, but I noticed that we broke out channels into a separate `channel.py`, worth considering here? I guess it depends how many projections we plan to add.

**Review by trbromley — COMMENTED** [review]
Thanks @Lucaman99, looking great! I have left some comments.
More generally, I was initially confused to see `Measure` as an operation within the circuit rather than something we return at the end. This makes sense after reading the ADR! But I was just wondering if we expect others to have this reaction, e.g., people trying to do things like `return qml.Measure(wires=range(2))`? Also, would we ever consider supporting, e.g., `return qml.probs(qml.Measure())`?

Apologies if this was all ironed out in the ADR, happy to defer to the decisions made there.

**Comment by Lucaman99** [comment]
Thanks for the review @trbromley, I'll get started on this right away!

File: pennylane/operation.py (hunk 5 above)
Comment (Lucaman99) [reply pennylane/operation.py:967]: Good question, I'm not familiar with the nuances of `Operation` vs `Operator`, I'm just going by the ADR.

File: pennylane/operation.py (hunk 7 above)
Comment (Lucaman99) [reply pennylane/operation.py:973]: I decided to keep it general as Nathan mentioned some possible applications of projections in QML circuits, so I thought maybe one may want to parametrize it at some point

File: pennylane/operation.py (hunk 10 above)
Comment (Lucaman99) [reply pennylane/operation.py:994]: Right, maybe we should remove it. Codecov is throwing a fit over us not testing it :joy:

File: pennylane/ops/qubit.py (hunk 14 above)
Comment (Lucaman99) [reply pennylane/ops/qubit.py:1885]: For sure!

File: pennylane/operation.py (hunk 13 above)
Comment (Lucaman99) [reply pennylane/operation.py:964]: Hmm, maybe, although we don't discuss other classes like `Tensor` and `Channel` here, so I'm wondering if we should make an exception for `Projection`?

File: pennylane/operation.py (hunk 1 above)
Comment (josh146) [reply pennylane/operation.py:988]: >does do_queue still apply to the new QNode/tape core?

No it doesn't :)

File: pennylane/operation.py (hunk 10 above)
Comment (josh146) [reply pennylane/operation.py:994]: > Channel does declare _kraus_matrices as an abstract method, but Operator doesn't have _matrix as an abstract method, so we're not consistent 🤔

Ah 🤔 

I think this is an oversight, needing to define `_kraus_matrices` shouldn't be an abstract method.

File: pennylane/operation.py (hunk 1 above)
Comment (Lucaman99) [reply pennylane/operation.py:988]: In that case, should this be removed @josh146? I was just trying to keep it consistent with the other classes.

Comment (josh146) [reply pennylane/operation.py:988]: Keep it in for now; when it comes time to rip out the old pennylane core, this can be removed en masse

File: pennylane/ops/projection.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    * Gradient recipe: None
+
+    Args:
+        wires (Sequence[int] or int): the wire(s) the operation acts on
+
+    .. UsageDetails::
+
+        It is important to note when using the `Measurement` operation that
+        one **applies** it to the wire one wishes to measure, and then returns
+        one of the PennyLane measurement functions.
```
Comment (ixfoduap) [inline pennylane/ops/projection.py:41]:
```suggestion
        The `Measurement` operation is applied to the wire we
        want to measure. The return of a circuit is then evaluated on
        the output state conditioned on the given measurement outcome.
        This will depend on the random outcome of the measurement,
        so expectation values should be obtained over several shots. 
```

File: pennylane/ops/projection.py [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+        For instance, consider the following, basic circuit:
```
Comment (ixfoduap) [inline pennylane/ops/projection.py:43]:
```suggestion
        For instance, consider the following circuit:
```

File: .github/CHANGELOG.md [hunk 18]
```diff
@@ -132,9 +152,7 @@ Thomas Bromley, Josh Izaac, Daniel Polatajko, Chase Roberts, Maria Schuld.
 
 This release contains contributions from (in alphabetical order):
 
-Thomas Bromley, Josh Izaac, Antal Száva.
-
-
+Thomas Bromley, Jack Ceroni, Josh Izaac, Daniel Polatajko, Chase Roberts, Maria Schuld
```
Comment (josh146) [inline .github/CHANGELOG.md:155]:
```suggestion
Thomas Bromley, Josh Izaac, Antal Száva.
```

File: doc/introduction/operations.rst [hunk 19]
```diff
@@ -142,6 +142,20 @@ Qubit observables
 
 :html:`</div>`
 
+.. _intro_ref_ops_measurements:
+
+Qubit measurements
+^^^^^^^^^^^^^^^^^^
```
Comment (josh146) [inline doc/introduction/operations.rst:148]: I feel like this will be a point of confusion to end-users - how will they distinguish between these mid-circuit projective measurements, and terminal measurements we already have?

Comment (josh146) [reply doc/introduction/operations.rst:148]: image

File: pennylane/ops/projection.py [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+"""
+
+import numpy as np
+
+from pennylane.operation import AnyWires, Projection
+
+
+class Measure(Projection):
+    r"""Measure(wires)
+    A measurement in the computational basis.
```
Comment (josh146) [inline pennylane/ops/projection.py:26]: I have to admit, reading further down, I'm getting confused myself! @trbromley, maybe you also have thoughts here.

So far, we have four measurement concepts/objects in PL:

1. `Projection`, an abstract base class that inherits from `Operation` (this makes sense to me).

2. `Measure`, which inherits from `Projection`.

3. Terminal measurement statistics, `qml.expval(..)`, `qml.var(..)`, etc. These have several roles:

    - They process the arguments, and return a `MeasurementProcess` instance, that contains all the information about the measurement.
    - They provide a user-interface to the user, without the user needing to create a measurement process object themselves. However, it _is_ possible to use a `MeasurementProcess` directly if you want, e.g., `return qml.MeasurementProcess(obs=qml.PauliZ(0))`

4. The `MeasurementProcess` class, which:

   - Contains information about the measurement (observables, eigenvalues, wires, etc).
   - Interacts with PennyLane's queuing system, to make sure that the process is added to the QNode.

To me, (1) makes sense.

However (2) confuses me - it's a specific implementation of the abstract `Projection` class, but it has a generic `Measure` name, which could also be confused with the existing (but separate) terminal measurement system!

Comment (josh146) [reply pennylane/ops/projection.py:26]: Something like the following, instead, makes a lot more sense from my point of view:

```python
@qml.qnode(dev)
def circuit():
    qml.Hadamard(wires=0)
    qml.CNOT(wires=[0, 1])
    qml.ProjectiveMeasure(projector=?, basis=qml.PauliX, wires=[0, 1])
    return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
```

* Having a more specific name (`ProjectiveMeasure`) reduces potential confusion, and emphasizes mid-circuit measurement.

* I would make `ProjectiveMeasure` a function, that returns a `MeasurementProcess` object with information about the projective measurement the device needs to undertake. That is, it should go in the `pennylane/tape/measure.py`, and modifications made to `MeasurementProcess` so that it can represent a projective measurement. This also avoids the need for having to subclass `Projector` every time you want to define a new projector; here, it is simply passed as a keyword argument.

I'm curious to get everyone's thoughts (@trbromley, @ixfoduap, @Lucaman99).

**Review by josh146 — COMMENTED** [review]
Thanks @Lucaman99! Your PR looks 💯, I just have a more conceptual question regarding the new classes

File: doc/introduction/operations.rst (hunk 19 above)
Comment (ixfoduap) [reply doc/introduction/operations.rst:148]: @josh146: This is a very good point. My personal opinion is that "Measurements" doesn't describe well the content inside this section of the documentation. Really it's more about circuit outputs, and how pennylane supports things like variances and probabilities. Additionally, I think that the name `Measure()` is perfect for the operation of measuring a wire mid circuit.

So my suggestion is that we consider changing the name "Measurements" in the documentation

Comment (josh146) [reply doc/introduction/operations.rst:149]: > So my suggestion is that we consider changing the name "Measurements" in the documentation

A good suggestion, my only hesitation is that it will be more invasive/work than just the title of the Measurements page unfortunately - the term 'measurements' (referring to post-processing terminal measurements) is used across the demos and API documentation as well.

File: pennylane/ops/projection.py (hunk 20 above)
Comment (ixfoduap) [reply pennylane/ops/projection.py:26]: Hi @josh146,

Thanks for the suggestions and thoughtfulness. I have to say, I'm not a big fan. I understand that there is some potential for confusion, but in my opinion this is minor. There are terminal measurement statistics, which are handled by ``MeasurementProcess``, and mid-circuit measurements that are handled by  ``Measure()``. From a user perspective, the main change that appears is that PennyLane now supports mid-circuit measurements, which are in a category of their own. Great!

I also prefer a UI with ``Measure()``, which is cleaner
```python
@qml.qnode(dev)
def circuit():
    qml.Hadamard(wires=0)
    qml.CNOT(wires=[0, 1])
    qml.Measure(wires=[0, 1])
    return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
```

Finally, we shouldn't forget everything that was discussed in the [ADR for conditional measurements](https://github.com/PennyLaneAI/adrs/blob/master/documents/041-conditional_operations.rst). I'm concerned that if we start deviating from that plan, we'll find ourselves having to deal with how these changes spread across the proposed architecture.

In short, I prefer to keep as is and just be extra careful in the documentation and tutorials to avoid potential confusions

File: .github/CHANGELOG.md [hunk 21]
```diff
@@ -45,6 +45,26 @@
   transforms.
   [(#1064)](https://github.com/PennyLaneAI/pennylane/pull/1064)
 
+* Adds `Projection` and `Measurement` classes to PennyLane.
```
Comment (trbromley) [inline .github/CHANGELOG.md:48]: Should `Measurement` be changed to `Measure` (+ a few cases below)

File: .github/CHANGELOG.md [hunk 22]
```diff
@@ -45,6 +45,26 @@
...
+  @qml.qnode(dev)
+  def circuit():
+      qml.Hadamard(wires=0)
+      qml.CNOT(wires=[0, 1])
+      qml.Measure(wires=[0, 1])
+
+      return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
+  ```
+  
+  Note that the `Measurement` operation must be **applied** to the wire, which
```
Comment (trbromley) [inline .github/CHANGELOG.md:65]: I'm not sure what we mean by applying it to the wire? Isn't that obvious?

File: doc/introduction/operations.rst (hunk 19 above)
Comment (trbromley) [reply doc/introduction/operations.rst:149]: As a compromise, could we call it "Qubit projections"?

File: pennylane/ops/projection.py (hunk 20 above)
Comment (trbromley) [reply pennylane/ops/projection.py:26]: Yes, my main worry is if a user tries to do: `return qml.Measure(wires=range(2))`. Indeed, this seems like a very natural thing to do and I would do it and be surprised that it didn't work. `Measure` seems to suggest extracting some classical information from the system. I think there will be less confusion with concepts like `Projection`, as it sounds less like extracting information from the system and more like performing an operation on it.

For the sake of following the ADR, could we just rename `class Measure(Projection):` to `class Project(Projection):`
Since it is projecting into the qubit computational basis, it can have a very general name. Other projections could be named, e.g., `ProjectPlusMinus`.

File: pennylane/ops/projection.py [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+                qml.Measure(wires=[0, 1])
+
+                return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
+
+        By running this circuit, we would get an output of the form:
+
+        >>> circuit()
+        array([1., 1.])
+
+        which corresponding to a measurement of :math:`|00\rangle`.
```
Comment (trbromley) [inline pennylane/ops/projection.py:64]: Do we want to add that if we were to run again, the result might be different. (e.g., sometimes get [-1, -1])

**Review by trbromley — COMMENTED** [review]
Thanks! I took a quick pass through and understand both perspectives of the discussion. Could we just rename `Measure` to dissuade users from thinking it's something that might be returned?

File: pennylane/ops/projection.py (hunk 20 above)
Comment (josh146) [reply pennylane/ops/projection.py:26]: > For the sake of following the ADR

I just want to emphasize that we do not need to adhere to an ADR simply because it was accepted! ADRs are not set in stone, and we can always revisit if we have new ideas or information. In fact, there have been quite a few cases in the past where reviewer comments during the implementation have led to (good) design changes.

If it becomes clear during the implementation of an already accepted ADR that there are

* unexpected problems, or
* new issues, or
* even if it becomes clear that there is a better way of designing/implementing,

it is perfectly fine to discuss and revise.

Regarding this comment thread specifically, my concern is less with the name and more in terms of the implementation; in particular, the need to create operation subclasses simply for defining new projectors.

With the new core being default, we have a plan to update the devices to understand the `MeasurementProcess` natively. The `MeasurementProcess` class already supports measurements defined via eigenvalues; would it not be still in scope to allow it, or a subclass of it, to represent projectors as well?

The

Comment (ixfoduap) [reply pennylane/ops/projection.py:26]: It seems like we need to have chat! Some friendly counterarguments:

- I mentioned the context of the ADR not because it means things are set in stone, but rather to make sure we keep all the pieces in mind. Ultimately we want to be able to use mid-circuit measurements to do several tasks. For example we want to do something like ``b = qml.Measure(wires=0)``, then post-porcess `b` and pass it to a subsequent operation. Also we want to interface with a new `qml.if()`. It took us month to see how all the pieces fit together and I worry that making one local change here takes us away from the global minimum
- I personally really like to think of `Measure()` as an operation: I can rotate the qubit, entangle it with a qubit, or measure it. Perhaps I need to understand `MeasurementProcess` better, but to me it seems like an overkill. Also, would it allows us to specify instructions in the simulators on how to update the state after a measurement has occured?
- The projection class was added as a suggestion from Nathan due to the potential usefulness of performing different kinds of projections in different contexts. We can definitely revisit this with new eyes
- I guess I would just like to see a clearer example of how we can use `MeasurementProcess` to support the same intended functionality that is done with this PR and that fits the bigger picture of supporting conditional operations (dynamical circuits) in PennyLane. This is unfortunately not very clear to me

**Comment by josh146** [comment]
Closing this, as mid circuit measurements are now part of PL :)
